### PR TITLE
[Module] Add remove mic check

### PR DESCRIPTION
**Pull request by XiaoliChan**
## Description
This module is used to check if it target vulnerable to `CVE-2019-1040` (aka `remove mic`)

In this module, it will redirect the ntlm negotiate function to the modified function (without mic), which means it will break the smb login when using this module, if target is not vulnerable, it will return `STATUS_INVALID_PARAMETER`

## Type of change
Please delete options that are not relevant.
- [√] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
nxc smb x.x.x.x/x -u valid_user -p valid_pass -M remove-mic

## Screenshots (if appropriate):
image

## Checklist:
- [√] I have ran Ruff against my changes (via poetry: `poetry run python -m ruff check . --preview`, use `--fix` to automatically fix what it can)
- [√] I have added or updated the tests/e2e_commands.txt file if necessary
- [√] New and existing e2e tests pass locally with my changes
- [√] My code follows the style guidelines of this project (should be covered by Ruff above)
- [×] If reliant on third party dependencies, such as Impacket, dploot, lsassy, etc, I have linked the relevant PRs in those projects
- [√] I have performed a self-review of my own code
- [√] I have commented my code, particularly in hard-to-understand areas
- [×] I have made corresponding changes to the documentation (PR here: https://github.com/Pennyw0rth/NetExec-Wiki)

### CODE REVIEW

**Comment by mpgn** [comment]
got this error on my side 🤔 

```
SMB         10.3.10.12      445    MEEREEN          [*] Windows Server 2016 Standard Evaluation 14393 x64 (name:MEEREEN) (domain:essos.local) (signing:True) (SMBv1:True)                                                                                                                                                               
SMB         10.3.10.12      445    MEEREEN          [+] sevenkingdoms.local\tywin.lannister:powerkingftw135 
[22:19:14] ERROR    Exception while calling proto_flow() on target 10.3.10.12: computeResponse() got an unexpected keyword argument                connection.py:174
                    'channel_binding_value'                                                                                                                         
                    ╭──────────────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────────────╮                  
                    │ /home/bonclay/NetExec/nxc/connection.py:166 in __init__                                                                    │                  
                    │                                                                                                                            │                  
                    │   163 │   │   self.logger.info(f"Socket info: host={self.host}, hostname={self.hostname},                                  │                  
                    │       kerberos={self.kerberos}, ipv6={self.is_ipv6}, link-local                                                            │                  
                    │       ipv6={self.is_link_local_ipv6}")                                                                                     │                  
                    │   164 │   │                                                                                                                │                  
                    │   165 │   │   try:                                                                                                         │                  
                    │ ❱ 166 │   │   │   self.proto_flow()                                                                                        │                  
                    │   167 │   │   except Exception as e:                                                                                       │                  
                    │   168 │   │   │   if "ERROR_DEPENDENT_SERVICES_RUNNING" in str(e):                                                         │                  
                    │   169 │   │   │   │   self.logger.error(f"Exception while calling proto_flow() on target                                   │                  
                    │       {target}: {e}")                                                                                                      │                  
                    │                                                                                                                            │                  
                    │ /home/bonclay/NetExec/nxc/connection.py:237 in proto_flow                                                                  │                  
                    │                                                                                                                            │                  
                    │   234 │   │   │   │   if hasattr(self.args, "module") and self.args.module:                                                │                  
                    │   235 │   │   │   │   │   self.load_modules()                                                                              │                  
                    │   236 │   │   │   │   │   self.logger.debug("Calling modules")                                                             │                  
                    │ ❱ 237 │   │   │   │   │   self.call_modules()                                                                              │                  
                    │   238 │   │   │   │   else:                                                                                                │                  
                    │   239 │   │   │   │   │   self.logger.debug("Calling command arguments")                                                   │                  
                    │   240 │   │   │   │   │   self.call_cmd_args()                                                                             │                  
                    │                                                                                                                            │                  
                    │ /home/bonclay/NetExec/nxc/connection.py:292 in call_modules                                                                │                  
                    │                                                                                                                            │                  
                    │   289 │   │   │                                                                                                            │                  
                    │   290 │   │   │   if hasattr(module, "on_login"):                                                                          │                  
                    │   291 │   │   │   │   self.logger.debug(f"Module {module.name} has on_login method")                                       │                  
                    │ ❱ 292 │   │   │   │   module.on_login(context, self)                                                                       │                  
                    │   293 │   │   │                                                                                                            │                  
                    │   294 │   │   │   if self.admin_privs and hasattr(module, "on_admin_login"):                                               │                  
                    │   295 │   │   │   │   self.logger.debug(f"Module {module.name} has on_admin_login method")                                 │                  
                    │                                                                                                                            │                  
                    │ /home/bonclay/NetExec/nxc/modules/remove-mic.py:46 in on_login                                                             │                  
                    │                                                                                                                            │                  
                    │    43 │   │   ntlm.computeResponseNTLMv2 = Modify_Func.mod_computeResponseNTLMv2                                           │                  
                    │    44 │   │   ntlm.getNTLMSSPType3 = Modify_Func.mod_getNTLMSSPType3                                                       │                  
                    │    45 │   │   try:                                                                                                         │                  
                    │ ❱  46 │   │   │   connection.conn.reconnect()                                                                              │                  
                    │    47 │   │   except SessionError as e:                                                                                    │                  
                    │    48 │   │   │   if e.getErrorCode() == nt_errors.STATUS_INVALID_PARAMETER:                                               │                  
                    │    49 │   │   │   │   context.log.info("Target is not vulnerable to CVE-2019-1040                                          │                  
                    │       (authentication was rejected)")                                                                                      │                  
                    │                                                                                                                            │                  
                    │ /home/bonclay/.cache/pypoetry/virtualenvs/netexec-YakB2KZD-py3.12/lib/python3.12/site-packages/impacket/smbconnection.py:9 │                  
                    │ 35 in reconnect                                                                                                            │                  
                    │                                                                                                                            │                  
                    │    932 │   │   if self._doKerberos is True:                                                                                │                  
                    │    933 │   │   │   self.kerberosLogin(userName, password, domain, lmhash, nthash, aesKey,                                  │                  
                    │        self._kdcHost, TGT, TGS, self._useCache)                                                                            │                  
                    │    934 │   │   else:                                                                                                       │                  
                    │ ❱  935 │   │   │   self.login(userName, password, domain, lmhash, nthash, self._ntlmFallback)                              │                  
                    │    936 │   │                                                                                                               │                  
                    │    937 │   │   return True                                                                                                 │                  
                    │    938                                                                                                                     │                  
                    │                                                                                                                            │                  
                    │ /home/bonclay/.cache/pypoetry/virtualenvs/netexec-YakB2KZD-py3.12/lib/python3.12/site-packages/impacket/smbconnection.py:2 │                  
                    │ 78 in login                                                                                                                │                  
                    │                                                                                                                            │                  
                    │    275 │   │   self._ntlmFallback = ntlmFallback                                                                           │                  
                    │    276 │   │   try:                                                                                                        │                  
                    │    277 │   │   │   if self.getDialect() == smb.SMB_DIALECT:                                                                │                  
                    │ ❱  278 │   │   │   │   return self._SMBConnection.login(user, password, domain, lmhash, nthash,                            │                  
                    │        ntlmFallback)                                                                                                       │                  
                    │    279 │   │   │   else:                                                                                                   │                  
                    │    280 │   │   │   │   return self._SMBConnection.login(user, password, domain, lmhash, nthash)                            │                  
                    │    281 │   │   except (smb.SessionError, smb3.SessionError) as e:                                                          │                  
                    │                                                                                                                            │                  
                    │ /home/bonclay/.cache/pypoetry/virtualenvs/netexec-YakB2KZD-py3.12/lib/python3.12/site-packages/impacket/smb.py:3504 in     │                  
                    │ login                                                                                                                      │                  
                    │                                                                                                                            │                  
                    │   3501 │   │                                                                                                               │                  
                    │   3502 │   │   if self._dialects_parameters['Capabilities'] & SMB.CAP_EXTENDED_SECURITY:                                   │                  
                    │   3503 │   │   │   try:                                                                                                    │                  
                    │ ❱ 3504 │   │   │   │   self.login_extended(user, password, domain, lmhash, nthash, use_ntlmv2 =                            │                  
                    │        True)                                                                                                               │                  
                    │   3505 │   │   │   except:                                                                                                 │                  
                    │   3506 │   │   │   │   # If the target OS is Windows 5.0 or Samba, let's try using NTLMv1                                  │                  
                    │   3507 │   │   │   │   if ntlm_fallback and                                                                                │                  
                    │        ((six.ensure_binary(self.get_server_lanman()).find(b'Windows 2000') != -1) or                                       │                  
                    │                                                                                                                            │                  
                    │ /home/bonclay/.cache/pypoetry/virtualenvs/netexec-YakB2KZD-py3.12/lib/python3.12/site-packages/impacket/smb.py:3412 in     │                  
                    │ login_extended                                                                                                             │                  
                    │                                                                                                                            │                  
                    │   3409 │   │   │   │   if len(version) >= 4:                                                                               │                  
                    │   3410 │   │   │   │      self.__server_os_major, self.__server_os_minor,                                                  │                  
                    │        self.__server_os_build = unpack('<BBH',version[:4])                                                                 │                  
                    │   3411 │   │   │                                                                                                           │                  
                    │ ❱ 3412 │   │   │   type3, exportedSessionKey = ntlm.getNTLMSSPType3(auth,                                                  │                  
                    │        respToken['ResponseToken'], user, password, domain, lmhash, nthash, use_ntlmv2 =                                    │                  
                    │        use_ntlmv2)                                                                                                         │                  
                    │   3413 │   │   │                                                                                                           │                  
                    │   3414 │   │   │   if exportedSessionKey is not None:                                                                      │                  
                    │   3415 │   │   │   │   self._SigningSessionKey = exportedSessionKey                                                        │                  
                    │                                                                                                                            │                  
                    │ /home/bonclay/NetExec/nxc/modules/remove-mic.py:141 in mod_getNTLMSSPType3                                                 │                  
                    │                                                                                                                            │                  
                    │   138 │   │                                                                                                                │                  
                    │   139 │   │   serverName = ntlmChallenge['TargetInfoFields']                                                               │                  
                    │   140 │   │                                                                                                                │                  
                    │ ❱ 141 │   │   ntResponse, lmResponse, sessionBaseKey =                                                                     │                  
                    │       ntlm.computeResponse(ntlmChallenge['flags'], ntlmChallenge['challenge'],                                             │                  
                    │   142 │   │   │   │   │   │   │   │   │   │   │   │   │   │   │   │   clientChallenge,                                     │                  
                    │       serverName, domain, user, password,                                                                                  │                  
                    │   143 │   │   │   │   │   │   │   │   │   │   │   │   │   │   │   │   lmhash, nthash,                                      │                  
                    │       use_ntlmv2, channel_binding_value= channel_binding_value)                                                            │                  
                    │   144                                                                                                                      │                  
                    ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯                  
                    TypeError: computeResponse() got an unexpected keyword argument 'channel_binding_value' `
```

**Comment by mpgn** [comment]
> > got this error on my side 🤔
> 
> You may have to update the `impacket` via `poetry`, because in the latest version of `impacket`, I saw @NeffIsBack add this variable
> 
> [fortra/impacket@ea27e8b](https://github.com/fortra/impacket/commit/ea27e8b2dfedf57370d2f65c5053a2b8eeb8ca9d)

all good 

This module is working well, but it should be noted that it fuck the ntlm auth for not vulnerable host therefore `multiple_hosts` should be set to false

image

File: nxc/modules/remove-mic.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+from impacket import nt_errors
+from impacket.smbconnection import SessionError
+
+
+class NXCModule:
+    name = "remove-mic"
+    description = "Check if host vulnerable to CVE-2019-1040"
+    supported_protocols = ["smb"]
+    opsec_safe = True
+    multiple_hosts = True
```
[the remark this answers is not in the document]
Comment (XiaoliChan) [reply nxc/modules/remove-mic.py]: Fixed

**Review by mpgn — APPROVED** [review]

**Comment by NeffIsBack** [comment]
It doesn't check for ntlmv1 right? It would be so so nice to have this included somehow

**Comment by XiaoliChan** [comment]
> It doesn't check for ntlmv1 right? It would be so so nice to have this included somehow

Yes, doesn't check for ntlmv1

**Pull request merged**